File: patterns/graceful-retry-actions/rules-js/graceful-retry-actions.js
const retryTimers = new Map();
const retryCounts = new Map();

rules.JSRule({
  name: "Configurable Multi-Device Retry",
  triggers: [triggers.ItemStateChangeTrigger("RetryTrigger", undefined, "ON")],
  execute: (event) => {
    const maxRetries = parseInt(items.getItem("RetryMaxAttempts").state);
    const initialInterval = parseInt(items.getItem("RetryInitialInterval").state);
    const maxInterval = parseInt(items.getItem("RetryMaxInterval").state);
    const altAction = items.getItem("RetryAlternativeAction").state;

    const devices = items.getItem("RetryDevices").members.map((i) => i.name);

    devices.forEach((device) => {
      retryCounts.set(device, 0);

      const attempt = () => {
        try {
          if (items.getItem(device).state !== "NULL") {
            items.getItem(device).sendCommand("ON");
            console.log(device + " command sent successfully!");
            if (retryTimers.has(device)) retryTimers.get(device).cancel();
          } else {
            throw new Error("Device " + device + " is offline");
          }
        } catch (ex) {
          const count = retryCounts.get(device) + 1;
          retryCounts.set(device, count);
          console.warn(device + " attempt #" + count + " failed: " + ex.message);

          if (count < maxRetries) {
            const nextInterval = Math.min(initialInterval * Math.pow(2, count), maxInterval);
            console.log("Next attempt for " + device + " in " + nextInterval + " seconds");
            retryTimers.get(device).reschedule(time.ZonedDateTime.now().plusSeconds(nextInterval));
          } else {
            console.error(device + " max retries reached!");
            const message = altAction ? device + ": " + altAction : device + " could not be switched ON!";
            actions.NotificationAction.sendNotification("admin@example.com", message);
          }
        }
      };

      const t = actions.ScriptExecution.createTimer(time.ZonedDateTime.now(), attempt);
      retryTimers.set(device, t);
    });
  }
});

File: patterns/graceful-retry-actions/rules-python/graceful_retry_actions.py
import threading

from openhab import rule, Registry, logger
from openhab.actions import NotificationAction
from openhab.triggers import ItemStateChangeTrigger

retry_timers = {}
retry_counts = {}


def attempt(device, max_retries, initial_interval, max_interval, alt_action):
    try:
        state = Registry.getItem(device).getState()
        if str(state) != "NULL":
            Registry.getItem(device).sendCommand("ON")
            logger.info(device + " command sent successfully!")
            if device in retry_timers:
                retry_timers[device].cancel()
        else:
            raise Exception("Device " + device + " is offline")
    except Exception as ex:
        count = retry_counts.get(device, 0) + 1
        retry_counts[device] = count
        logger.warn("{} attempt #{} failed: {}".format(device, count, ex))

        if count < max_retries:
            next_interval = min(initial_interval * (2 ** count), max_interval)
            logger.info("Next attempt for {} in {} seconds".format(device, next_interval))
            t = threading.Timer(
                next_interval, attempt, args=(device, max_retries, initial_interval, max_interval, alt_action)
            )
            t.start()
            retry_timers[device] = t
        else:
            logger.error(device + " max retries reached!")
            message = (device + ": " + alt_action) if alt_action else (device + " could not be switched ON!")
            NotificationAction.sendNotification("admin@example.com", message)


@rule(triggers=[ItemStateChangeTrigger("RetryTrigger", state="ON")])
class ConfigurableMultiDeviceRetry:
    def execute(self, module, input):
        max_retries = int(str(Registry.getItem("RetryMaxAttempts").getState()))
        initial_interval = int(str(Registry.getItem("RetryInitialInterval").getState()))
        max_interval = int(str(Registry.getItem("RetryMaxInterval").getState()))
        alt_action = str(Registry.getItem("RetryAlternativeAction").getState())

        devices = [i.getName() for i in Registry.getItem("RetryDevices").getAllMembers()]

        for device in devices:
            retry_counts[device] = 0
            attempt(device, max_retries, initial_interval, max_interval, alt_action)
